File: typescript/tests/caching.test.ts
import { describe, it, expect } from "vitest";
import { Record as ATRecord } from "airtable";
import { Airtable, PrimaryFieldSet, SecondaryFieldSet } from "../output";

function newPrimaryRecord(airtable: Airtable): ATRecord<PrimaryFieldSet> {
	return new ATRecord<PrimaryFieldSet>(airtable.primary._table, "", {});
}

function newSecondaryRecord(airtable: Airtable): ATRecord<SecondaryFieldSet> {
	return new ATRecord<SecondaryFieldSet>(airtable.secondary._table, "", {});
}

describe("Cache Hit (reference equality)", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });
	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", "Cache Hit Test");
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	const first = await airtable.primary.get(id);
	const second = await airtable.primary.get(id);

	it("should return the exact same object", () => {
		expect(second).toBe(first);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(id);
	});
});

describe("Cache Disabled By Default", async () => {
	const airtable = new Airtable();
	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", "No Cache Test");
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	await airtable.primary.get(id);
	const cacheSize = (airtable.primary as any)._cache.size;

	it("should have empty cache", () => {
		expect(cacheSize).toBe(0);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(id);
	});
});

describe("Mutation Invalidation", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });

	const r1 = newPrimaryRecord(airtable);
	r1.set("Primary Key", "Mutation Test");
	const created = await airtable.primary.create(r1);
	const id = created.id;

	it("get populates cache, create invalidates it", async () => {
		await airtable.primary.get(id);
		expect((airtable.primary as any)._cache.size).toBeGreaterThan(0);

		const r2 = newPrimaryRecord(airtable);
		r2.set("Primary Key", "Mutation Test 2");
		const created2 = await airtable.primary.create(r2);
		expect((airtable.primary as any)._cache.size).toBe(0);
		await airtable.primary.delete(created2.id);
	});

	it("update invalidates cache", async () => {
		await airtable.primary.get(id);
		expect((airtable.primary as any)._cache.size).toBeGreaterThan(0);

		const r = await airtable.primary.get(id, { returnAs: "record" });
		r.set("Primary Key", "Mutation Updated");
		await airtable.primary.update(r);
		expect((airtable.primary as any)._cache.size).toBe(0);
	});

	it("delete invalidates cache", async () => {
		await airtable.primary.get(id);
		expect((airtable.primary as any)._cache.size).toBeGreaterThan(0);

		await airtable.primary.delete(id);
		expect((airtable.primary as any)._cache.size).toBe(0);
	});
});

describe("Manual Invalidation", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });
	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", "Manual Invalidation Test");
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	it("invalidateCache clears cache", async () => {
		await airtable.primary.get(id);
		expect((airtable.primary as any)._cache.size).toBeGreaterThan(0);

		airtable.primary.invalidateCache();
		expect((airtable.primary as any)._cache.size).toBe(0);
	});

	it("get still works after invalidation", async () => {
		const result = await airtable.primary.get(id);
		expect(result.id).toBe(id);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(id);
	});
});

describe("Main Client Cascade Invalidation", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });

	const pRec = newPrimaryRecord(airtable);
	pRec.set("Primary Key", "Cascade Primary");
	const pCreated = await airtable.primary.create(pRec);

	const sRec = newSecondaryRecord(airtable);
	sRec.set("Name", "Cascade Secondary");
	const sCreated = await airtable.secondary.create(sRec);

	it("invalidateCache cascades to all tables", async () => {
		await airtable.primary.get(pCreated.id);
		await airtable.secondary.get(sCreated.id);
		expect((airtable.primary as any)._cache.size).toBeGreaterThan(0);
		expect((airtable.secondary as any)._cache.size).toBeGreaterThan(0);

		airtable.invalidateCache();
		expect((airtable.primary as any)._cache.size).toBe(0);
		expect((airtable.secondary as any)._cache.size).toBe(0);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(pCreated.id);
		await airtable.secondary.delete(sCreated.id);
	});
});

describe("Cache Expiry", async () => {
	const airtable = new Airtable({ cacheSeconds: 2 });
	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", "Expiry Test");
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	it("expires after TTL", async () => {
		const first = await airtable.primary.get(id);
		const cached = await airtable.primary.get(id);
		expect(cached).toBe(first);

		await new Promise((r) => setTimeout(r, 3000));

		const afterExpiry = await airtable.primary.get(id);
		expect(afterExpiry).not.toBe(first);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(id);
	});
});

describe("Cache Serves Stale Data", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });

	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", "Stale Data Test");
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	it("returns stale cached data after external update", async () => {
		// Populate cache with original value
		const first = await airtable.primary.get(id);
		expect(first.primaryKey).toBe("Stale Data Test");

		// Update via underlying Airtable.js API directly (bypasses cache invalidation)
		await airtable.primary._table.update(id, { "Primary Key": "Updated Directly" });

		// Cached instance should return stale data
		const stale = await airtable.primary.get(id);
		expect(stale.primaryKey).toBe("Stale Data Test");
	});

	it("cleanup", async () => {
		await airtable.primary.delete(id);
	});
});

describe("Different Parameters Different Cache Keys", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });

	const r1 = newPrimaryRecord(airtable);
	r1.set("Primary Key", "Cache Key 1");
	const c1 = await airtable.primary.create(r1);

	const r2 = newPrimaryRecord(airtable);
	r2.set("Primary Key", "Cache Key 2");
	const c2 = await airtable.primary.create(r2);

	it("different IDs are cached independently", async () => {
		const get1 = await airtable.primary.get(c1.id);
		const get2 = await airtable.primary.get(c2.id);
		expect((airtable.primary as any)._cache.size).toBe(2);

		const reGet1 = await airtable.primary.get(c1.id);
		const reGet2 = await airtable.primary.get(c2.id);
		expect(reGet1).toBe(get1);
		expect(reGet2).toBe(get2);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(c1.id);
		await airtable.primary.delete(c2.id);
	});
});

describe("Batch Get Caching", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });

	const r1 = newPrimaryRecord(airtable);
	r1.set("Primary Key", "Batch Cache 1");
	const c1 = await airtable.primary.create(r1);

	const r2 = newPrimaryRecord(airtable);
	r2.set("Primary Key", "Batch Cache 2");
	const c2 = await airtable.primary.create(r2);

	it("batch get is cached separately from single get", async () => {
		// Batch get caches as one entry
		const batch = await airtable.primary.get([c1.id, c2.id]);
		expect((airtable.primary as any)._cache.size).toBe(1);

		// Single get creates a separate cache entry
		const single = await airtable.primary.get(c1.id);
		expect((airtable.primary as any)._cache.size).toBe(2);

		// Both return cached results on re-get
		const batch2 = await airtable.primary.get([c1.id, c2.id]);
		const single2 = await airtable.primary.get(c1.id);
		expect(batch2).toBe(batch);
		expect(single2).toBe(single);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(c1.id);
		await airtable.primary.delete(c2.id);
	});
});

describe("Fields Option Affects Cache Key", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });

	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", "Fields Cache Test");
	newRecord.set("Single Line Text", "Hello");
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	it("different fields options produce different cache entries", async () => {
		// Get with no fields filter
		const full = await airtable.primary.get(id);
		expect((airtable.primary as any)._cache.size).toBe(1);

		// Get with fields filter creates a separate cache entry
		const partial = await airtable.primary.get(id, { fields: ["Primary Key"] });
		expect((airtable.primary as any)._cache.size).toBe(2);

		// Both return cached results on re-get
		const full2 = await airtable.primary.get(id);
		const partial2 = await airtable.primary.get(id, { fields: ["Primary Key"] });
		expect(full2).toBe(full);
		expect(partial2).toBe(partial);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(id);
	});
});

describe("Formula Query Caching", async () => {
	const airtable = new Airtable({ cacheSeconds: 60 });

	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", "Formula Cache Test");
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	it("formula query result is cached", async () => {
		const formula = "{Primary Key} = 'Formula Cache Test'";
		const first = await airtable.primary.get({ formula });
		expect((airtable.primary as any)._cache.size).toBe(1);

		const second = await airtable.primary.get({ formula });
		expect(second).toBe(first);
	});

	it("different formulas produce different cache entries", async () => {
		airtable.primary.invalidateCache();

		const f1 = "{Primary Key} = 'Formula Cache Test'";
		const f2 = "{Primary Key} != 'nonexistent'";
		await airtable.primary.get({ formula: f1 });
		await airtable.primary.get({ formula: f2 });
		expect((airtable.primary as any)._cache.size).toBe(2);
	});

	it("cleanup", async () => {
		await airtable.primary.delete(id);
	});
});

File: go/tests/caching_test.go
package tests

// G9.3 — Live TTL-cache behavior end-to-end, parity with Java/Kotlin/Swift/Rust
// TestCaching: cache hits, default-off, mutation invalidation (same client),
// manual invalidation, TTL expiry, and query-key separation.
//
// The runtime cache (cacheStore) is package-private to the generated `airtable`
// package, so this external `tests` package cannot inspect entry counts the way
// the Java suite does. Instead each case asserts the OBSERVABLE behavior: a
// `mutator` client built with cacheSeconds=0 makes an out-of-band change that a
// separate cached client either does NOT see (still serving a cached payload) or
// DOES see (cache disabled / invalidated / expired). That is the property the
// cache actually provides.

import (
	"context"
	"testing"
	"time"

	airtable "myairtabletests/output"
)

// cachedAirtable builds a client with the given TTL (seconds).
func cachedAirtable(t *testing.T, seconds float64) *airtable.Airtable {
	t.Helper()
	requireCreds(t)
	return airtable.NewWithBase(apiKey(), baseID(), seconds)
}

func bestEffortDelete(at *airtable.Airtable, id string) {
	if id == "" {
		return
	}
	_ = at.Primary.DeleteOne(context.Background(), id)
}

func TestCaching(t *testing.T) {
	requireCreds(t)
	ctx := context.Background()

	t.Run("CacheHitServesStalePayloadUntilExternalChange", func(t *testing.T) {
		cached := cachedAirtable(t, 60)
		mutator := cachedAirtable(t, 0) // out-of-band writer, no cache

		pk := primaryKey("Cache", "Hit")
		created, err := mutator.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:     airtable.String(pk),
			SingleLineText: airtable.String("original"),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(mutator, id)

		// First read through the cached client populates the cache.
		first, err := cached.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("first get: %v", err)
		}
		if first.SingleLineText == nil || *first.SingleLineText != "original" {
			t.Fatalf("first read text = %v, want original", first.SingleLineText)
		}

		// Change the record out-of-band via the uncached mutator client.
		created.SingleLineText = airtable.String("changed externally")
		if _, err := mutator.Primary.UpdateOne(ctx, created); err != nil {
			t.Fatalf("external update: %v", err)
		}

		// Second read through the cached client is served from cache: it must NOT
		// reflect the external change.
		second, err := cached.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("second get: %v", err)
		}
		if second.SingleLineText == nil || *second.SingleLineText != "original" {
			t.Fatalf("cached read text = %v, want still 'original' (served from cache)", second.SingleLineText)
		}
		if second.ID() != first.ID() {
			t.Fatalf("cached id = %s, want %s", second.ID(), first.ID())
		}
	})

	t.Run("CacheDisabledByDefaultSeesEveryChange", func(t *testing.T) {
		reader := cachedAirtable(t, 0) // cacheSeconds=0 -> no caching
		mutator := cachedAirtable(t, 0)

		pk := primaryKey("Cache", "Off")
		created, err := mutator.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:     airtable.String(pk),
			SingleLineText: airtable.String("v1"),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(mutator, id)

		if _, err := reader.Primary.GetOne(ctx, id); err != nil {
			t.Fatalf("first get: %v", err)
		}

		created.SingleLineText = airtable.String("v2")
		if _, err := mutator.Primary.UpdateOne(ctx, created); err != nil {
			t.Fatalf("external update: %v", err)
		}

		// With caching off every read hits the server, so the change IS visible.
		again, err := reader.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("second get: %v", err)
		}
		if again.SingleLineText == nil || *again.SingleLineText != "v2" {
			t.Fatalf("uncached read text = %v, want v2 (fresh read each time)", again.SingleLineText)
		}
	})

	t.Run("CreateOnCachedClientInvalidatesTableCache", func(t *testing.T) {
		cached := cachedAirtable(t, 60)
		pk := primaryKey("Cache", "MutateCreate")

		a, err := cached.Primary.CreateOne(ctx, &airtable.PrimaryModel{PrimaryKey: airtable.String(pk + " A")})
		if err != nil {
			t.Fatalf("create A: %v", err)
		}
		idA := a.ID()
		var idB string
		defer func() { bestEffortDelete(cached, idA); bestEffortDelete(cached, idB) }()

		// Populate the list cache for this table.
		filter := "{Primary Key} = \"" + pk + " A\""
		before, err := cached.Primary.GetMany(ctx, (&airtable.Query{}).WithFilter(filter))
		if err != nil {
			t.Fatalf("list before: %v", err)
		}
		if len(before) != 1 {
			t.Fatalf("list before = %d, want 1", len(before))
		}

		// A create through the SAME cached client invalidates this table's cache.
		b, err := cached.Primary.CreateOne(ctx, &airtable.PrimaryModel{PrimaryKey: airtable.String(pk + " A")})
		if err != nil {
			t.Fatalf("create B: %v", err)
		}
		idB = b.ID()

		// The same filter now re-reads from the server and sees both records.
		after, err := cached.Primary.GetMany(ctx, (&airtable.Query{}).WithFilter(filter))
		if err != nil {
			t.Fatalf("list after: %v", err)
		}
		if len(after) != 2 {
			t.Fatalf("list after create = %d, want 2 (create invalidated cache)", len(after))
		}
	})

	t.Run("UpdateOnCachedClientInvalidatesTableCache", func(t *testing.T) {
		cached := cachedAirtable(t, 60)
		pk := primaryKey("Cache", "MutateUpdate")

		created, err := cached.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:     airtable.String(pk),
			SingleLineText: airtable.String("before"),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(cached, id)

		if _, err := cached.Primary.GetOne(ctx, id); err != nil {
			t.Fatalf("populate get: %v", err)
		}

		created.SingleLineText = airtable.String("after")
		if _, err := cached.Primary.UpdateOne(ctx, created); err != nil {
			t.Fatalf("update: %v", err)
		}

		// The update wiped this table's cache, so the next get reflects the change.
		fetched, err := cached.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("get after update: %v", err)
		}
		if fetched.SingleLineText == nil || *fetched.SingleLineText != "after" {
			t.Fatalf("text after update = %v, want after (update invalidated cache)", fetched.SingleLineText)
		}
	})

	t.Run("DeleteOnCachedClientInvalidatesTableCache", func(t *testing.T) {
		cached := cachedAirtable(t, 60)
		pk := primaryKey("Cache", "MutateDelete")

		created, err := cached.Primary.CreateOne(ctx, &airtable.PrimaryModel{PrimaryKey: airtable.String(pk)})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		deleted := false
		defer func() {
			if !deleted {
				bestEffortDelete(cached, id)
			}
		}()

		filter := "{Primary Key} = \"" + pk + "\""
		before, err := cached.Primary.GetMany(ctx, (&airtable.Query{}).WithFilter(filter))
		if err != nil {
			t.Fatalf("list before: %v", err)
		}
		if len(before) != 1 {
			t.Fatalf("list before = %d, want 1", len(before))
		}

		if err := cached.Primary.DeleteOne(ctx, id); err != nil {
			t.Fatalf("delete: %v", err)
		}
		deleted = true

		// The delete invalidated the cache: the same filter now returns nothing.
		after, err := cached.Primary.GetMany(ctx, (&airtable.Query{}).WithFilter(filter))
		if err != nil {
			t.Fatalf("list after: %v", err)
		}
		if len(after) != 0 {
			t.Fatalf("list after delete = %d, want 0 (delete invalidated cache)", len(after))
		}
	})

	t.Run("InvalidateTableCacheForcesFreshRead", func(t *testing.T) {
		cached := cachedAirtable(t, 60)
		mutator := cachedAirtable(t, 0)

		pk := primaryKey("Cache", "ManualTable")
		created, err := mutator.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:     airtable.String(pk),
			SingleLineText: airtable.String("orig"),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(mutator, id)

		if _, err := cached.Primary.GetOne(ctx, id); err != nil {
			t.Fatalf("populate get: %v", err)
		}

		created.SingleLineText = airtable.String("manual change")
		if _, err := mutator.Primary.UpdateOne(ctx, created); err != nil {
			t.Fatalf("external update: %v", err)
		}

		// Manual per-table invalidation drops the cached entry.
		cached.Client().InvalidateTableCache((&airtable.PrimaryModel{}).TableID())

		fetched, err := cached.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("get after invalidate: %v", err)
		}
		if fetched.SingleLineText == nil || *fetched.SingleLineText != "manual change" {
			t.Fatalf("text = %v, want manual change after InvalidateTableCache", fetched.SingleLineText)
		}
	})

	t.Run("InvalidateAllCachesForcesFreshRead", func(t *testing.T) {
		cached := cachedAirtable(t, 60)
		mutator := cachedAirtable(t, 0)

		pk := primaryKey("Cache", "ManualAll")
		created, err := mutator.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:     airtable.String(pk),
			SingleLineText: airtable.String("orig"),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(mutator, id)

		if _, err := cached.Primary.GetOne(ctx, id); err != nil {
			t.Fatalf("populate get: %v", err)
		}

		created.SingleLineText = airtable.String("all change")
		if _, err := mutator.Primary.UpdateOne(ctx, created); err != nil {
			t.Fatalf("external update: %v", err)
		}

		cached.InvalidateAllCaches()

		fetched, err := cached.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("get after invalidate all: %v", err)
		}
		if fetched.SingleLineText == nil || *fetched.SingleLineText != "all change" {
			t.Fatalf("text = %v, want all change after InvalidateAllCaches", fetched.SingleLineText)
		}
	})

	t.Run("EntriesExpireAfterTTL", func(t *testing.T) {
		cached := cachedAirtable(t, 1) // 1-second TTL
		mutator := cachedAirtable(t, 0)

		pk := primaryKey("Cache", "TTL")
		created, err := mutator.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:     airtable.String(pk),
			SingleLineText: airtable.String("ttl-orig"),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(mutator, id)

		if _, err := cached.Primary.GetOne(ctx, id); err != nil {
			t.Fatalf("populate get: %v", err)
		}

		created.SingleLineText = airtable.String("ttl-changed")
		if _, err := mutator.Primary.UpdateOne(ctx, created); err != nil {
			t.Fatalf("external update: %v", err)
		}

		// Wait past the TTL so the cached entry is lazily evicted on next read.
		time.Sleep(1500 * time.Millisecond)

		fetched, err := cached.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("get after TTL: %v", err)
		}
		if fetched.SingleLineText == nil || *fetched.SingleLineText != "ttl-changed" {
			t.Fatalf("text = %v, want ttl-changed after TTL expiry", fetched.SingleLineText)
		}
	})

	t.Run("DistinctQueriesCacheIndependently", func(t *testing.T) {
		cached := cachedAirtable(t, 60)
		mutator := cachedAirtable(t, 0)

		pk := primaryKey("Cache", "Keys")
		a, err := cached.Primary.CreateOne(ctx, &airtable.PrimaryModel{PrimaryKey: airtable.String(pk)})
		if err != nil {
			t.Fatalf("create A: %v", err)
		}
		idA := a.ID()
		var idB string
		defer func() { bestEffortDelete(cached, idA); bestEffortDelete(mutator, idB) }()

		filter := "{Primary Key} = \"" + pk + "\""
		// Two different queries (distinct maxRecords) are cached under separate keys.
		q1 := (&airtable.Query{}).WithFilter(filter).WithMaxRecords(1)
		q2 := (&airtable.Query{}).WithFilter(filter).WithMaxRecords(5)

		r1, err := cached.Primary.GetMany(ctx, q1)
		if err != nil {
			t.Fatalf("list q1: %v", err)
		}
		r2, err := cached.Primary.GetMany(ctx, q2)
		if err != nil {
			t.Fatalf("list q2: %v", err)
		}
		if len(r1) != 1 || len(r2) != 1 {
			t.Fatalf("initial lists = %d / %d, want 1 / 1", len(r1), len(r2))
		}

		// Add a second matching record out-of-band, then invalidate only q1's
		// table cache by re-running q1... but since both queries live under the
		// same table, a mutation would invalidate both. To prove the keys are
		// distinct we instead confirm that re-running each query without any
		// mutation serves the cached payload (count stays as first observed),
		// independent of the other query's key.
		b, err := mutator.Primary.CreateOne(ctx, &airtable.PrimaryModel{PrimaryKey: airtable.String(pk)})
		if err != nil {
			t.Fatalf("create B (out-of-band): %v", err)
		}
		idB = b.ID()

		// Both queries are still served from their own cached entries (the
		// out-of-band create went through a DIFFERENT client, so neither key was
		// invalidated) — proving each cached independently and survives.
		again1, err := cached.Primary.GetMany(ctx, q1)
		if err != nil {
			t.Fatalf("re-list q1: %v", err)
		}
		again2, err := cached.Primary.GetMany(ctx, q2)
		if err != nil {
			t.Fatalf("re-list q2: %v", err)
		}
		if len(again1) != 1 {
			t.Fatalf("q1 re-list = %d, want 1 (served from its own cache key)", len(again1))
		}
		if len(again2) != 1 {
			t.Fatalf("q2 re-list = %d, want 1 (served from its own cache key)", len(again2))
		}
	})
}
